src/rgw/rgw_tools.cc

```diff
@@ -136,6 +138,10 @@ int rgw_rados_operate(librados::IoCtx& ioctx, const std::string& oid,
     }
     return -ec.value();
   }
+  // work on asio threads should be asynchronous, so warn when they block
+  if (is_asio_thread) {
+    dout(20) << "WARNING: blocking librados call" << dendl;
```

Comment:
> we can consider asserting in these cases.

it will take a while before we get everything under process_request() converted, and even then it will be hard to tell for sure without 100% test coverage - so i'd really prefer not to crash in a customer environment, and instead just log warnings that we can scan for in teuthology

> Also, maybe we can decipher some more generic way to identify blocking calls in these cases (e.g., lower level hook on a syscall -- when running in dev enviornment).

that's a really cool idea.. though boost::asio::io_context itself uses blocking operations to manage its work queue, so we'd need some way to filter those out. and we'll still run some background threads like gc/lifecycle that i wasn't planning to convert until much later

i do agree that this thread_local variable isn't a great solution, especially when it can't cover cls client calls that use IoCtx.operate() directly. we might consider adding some kind of async-only flag to IoCtx that causes it to fail any synchronous calls with -EWOULDBLOCK?